test(rating): add unit test for rating component

frontendwizard opened the pull request:
Adding tests for the rating component. You mentioned on the last livestream that this would be welcome so I took a stab at it.

There's a couple of findings on this test that caught me off guard and I'd be more than happy to hear what y'all think of it.

1. I was forced into using an empty `waitFor` on line 27 for `react-query` to catch up and perform the update on the database. I feel like it's a similar issue that's mentioned on the `setupTests` file.
2. Because the component receives the `listItem` through props and doesn't read it from the database, I had to rerender it with the updated list item from the database. If we passed just the `bookId` and then make use of `useListItem` to read from the db than this would not be needed. I experimented with and it's a rather simple change. I just choose not to because it'd also change the component API.

kentcdodds commented on src/components/__tests__/rating.js:
```diff
@@ -0,0 +1,31 @@
+
+  userEvent.click(firstStar)
+
+  // needed because of react-query
+  await waitFor(() => {})
+  const updatedListItem = await listItemsDB.read(listItem.id)
+  rerender(<Rating listItem={updatedListItem} />)
+  expect(screen.getByLabelText('1 star')).toBeChecked()
```
This one's actually a bit tricky. What we really care about is that the right radio gets selected and the database is updated (eventually).

Unfortunately, `waitFor` doesn't handle async stuff within the callback function (this may be a useful feature to build in though eventually).

So here's what I came up with:

```suggestion
  expect(screen.getByLabelText('1 star')).toBeChecked()
  let updatedListItem
  do {
    updatedListItem = await listItemsDB.read(listItem.id)
  } while (updateListItem.rating !== 1)
```

I don't love it, but I think it's the best way to get the confidence in the things we want.

kentcdodds reviewed (commented):
Looks awesome! I have a suggestion. Thank you :)

frontendwizard commented:
love the suggestion, definitely a better way then the empty `waitFor`.
Just fixed a typo on the while statement and added the rerender again to check wether the right option is selected. If you don't rerender, the component won't update itself because it doesn't read from the db. We could change the component to read from the db instead of receiving from props, but that'd change the api.

frontendwizard commented:
One thought though, is this test going to be stuck in an infinite loop if the database doesn't update? Is this ok?

kentcdodds commented:
Because the loop is async (thanks to the use of the `await` keyword), we don't need to worry about an infinite loop. It'll simply timeout (at least it should.... did you test that out?).

We definitely don't want the component to read directly from the db because the db is for testing purposes only anyway. However, the component *should* update thanks to react-query.

> If you don't rerender, the component won't update itself

I'm not sure what you mean by this. What makes you think it wouldn't update itself? The assertion that the 1 star rating is applied happens both before and after the rerender call, so it appears there's no difference in the output anyway. Isn't this just an unnecessary rerender?

kentcdodds commented:
Also, thanks for working on this.

frontendwizard commented:
I'm sorry, when I mean reading from the db, I actually mean reading from react-query cache. The db is just the server replacement for the tests.
my argument is that he's not reading from the react-query cache, but rather the parent is. that way, even if after the mutation happens, the component doesn't know about it because it's not using the `useListItem` hook, it's receiving through props from the parent, who's the one reading from the cache.

frontendwizard commented:
for this component to update itself after the mutation, it needs to read the list item through the `useListItem` hook and receive only the `bookId` from the props. I can make that happen if you agree.

frontendwizard commented:
This is how it'd look like with the component getting `listItem` from the hook:

```
// test file
test('it updates the rating', async () => {
  const {unmount} = await renderRating()
  const firstStar = screen.getByLabelText('1 star')

  act(() => userEvent.click(firstStar))

  waitFor(() => expect(screen.getByLabelText('1 star')).toBeChecked())
  // need to unmount because the react-query cache is cleaned up before the component
  // is unmounted, causing the `useListItem` hook to return null, breaking the component
  unmount()
}
// component
function Rating({bookId}) {
  const listItem = useListItem(bookId)
```

kentcdodds commented on src/components/__tests__/rating.js:
```diff
@@ -0,0 +1,33 @@
+
+  userEvent.click(firstStar)
+
+  let updatedListItem
+  do {
+    updatedListItem = await listItemsDB.read(listItem.id)
+  } while (updatedListItem.rating !== 1)
```
Once [this is merged](https://github.com/testing-library/dom-testing-library/pull/782) (in the next day or so), then we should be able to upgrade and make this much better:

```suggestion
  await waitFor(async () => {
    const updatedListItem = await listItemsDB.read(listItem.id)
    expect(updatedListItem.rating).toBe(1)
  })
```

kentcdodds commented on src/components/__tests__/rating.js:
```diff
@@ -0,0 +1,56 @@
+import React from 'react'
+import {render, screen, loginAsUser} from 'test/app-test-utils'
```
```suggestion
import {render, screen, waitFor, loginAsUser} from 'test/app-test-utils'
```

kentcdodds reviewed (approved):
Super 👍👍

kentcdodds commented:
@all-contributors please add @frontendwizard for tests
